text/0000-deprecate-string-based-actions.md

```diff
@@ -0,0 +1,100 @@
+- Start Date: 2020-05-23
+- Relevant Team(s): Ember.js
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# Deprecate String Based Actions
+
+## Motivation
+
+> In Ember 1.13 we introduced [closure actions](https://github.com/emberjs/rfcs/blob/00ac2685c86f27d41547012903f485a4ef338d27/active/0000-improved-actions.md), which have been recommended over string based actions. This officially deprecates string based actions in favor of closure actions,
+which provide many benefits such as
+
+1. simpler passing through components
+2. less need for an actions hash
+3. ability to return a value
+4. typing in TypeScript
+5. more in line with the Octane programming model
+
+## Detailed Design
+
+We will deprecate the following:
+
+1. The `send` api.
+2. The use of a string parameter to the `action` helper.
+3. The use of a string parameter to the `action` modifier.
+
+The `send` api may currently be used in components, controllers, and routes.
+It is used to target the `actions` hash. In calling the `actions` hash of the
+same object, `send` can be replaced by a function call once the method is
+moved outside the `actions` hash.
+
+For cases where `send` is used to go from a controller to a route or from
+a route to a parent route, no direct replacement for this "action bubbling"
+behavior will be provided.
+Instead, users are recommended to inject the router service.
+
+## Transition Path
+
+Users will transition to closure actions.
+
+From:
+
+```js
+export default Component.extend({
+    someMethod() {
+        this.send('actionName');
+    },
+
+    actions: {
+        actionName() {
+            // do something
+        }
+    }
+});
+```
+
+To:
+
+```js
+export default Component.extend({
+    someMethod() {
+        this.actionName();
+    },
+
+    actionName() {
+        // do something
+    }
+});
+```
+
+From:
+
+```hbs
+<button {{action 'actionName'}}>Action Label</button>
+```
+
+To:
+
+```hbs
+<button {{action this.actionName}}>Action Label</button>
+```
+
+## How We Teach This
+
+> We already teach closure actions exclusively in the guides. Other than a new
+deprecation guide, no additional material should be necessary.
+
+## Drawbacks
+
+> Some older applications have thousands of string based actions throughout their codebases.
+We should consider writing a codemod to ease the transition.
+
+## Alternatives
+
+> Closure actions are now the standard in Ember, so other alternatives have not been
+seriously considered.
+
+## Unresolved questions
```

Comment:
A couple other questions I have:

1. Can the `target` property of core classes be deprecated also? I didn't dig deep enough, so I'm not sure what else it's used for
1. Can the `ActionHandler` mixin be deprecated entirely? It only handles the `send` method and merges action hashes, so in theory it could be deprecated. (Although, since it's private API, I'm not sure it's worth signaling that)

Reply:
> Can the target property of core classes be deprecated also? I didn't dig deep enough, so I'm not sure what else it's used for.

Not sure either. I can add it to the list of unresolved questions.

> Can the ActionHandler mixin be deprecated entirely? It only handles the send method and merges action hashes, so in theory it could be deprecated. (Although, since it's private API, I'm not sure it's worth signaling that)

I don't think it needs a separate deprecation, and throwing two deprecations when using `send` is noisy. It should be removable once `send` is removed from Ember.

Reply:
Since the mixin is on the global, it might still be good to explicitly issue a deprecation if you access it. The internal usages in Ember can be updated to avoid triggering that secondary / knock on deprecation.